### PR TITLE
[Fix]: Storage Input/Output types

### PR DESCRIPTION
**Pull request by ashika112**
#### Description of changes
 Refactoring the input and output type to export both withKey and withPath types to customers

#### Description of how you validated changes
- Unit test
- Manual test
- Integ test

#### Checklist
- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/providers/s3/apis/server/getUrl.ts [hunk 1]
```diff
@@ -8,65 +8,61 @@ import {
...
+ * @returns Presigned URL and timestamp when the URL MAY expire.
+ * @throws service: `S3Exception` - thrown when checking for existence of the object
+ * @throws validation: `StorageValidationErrorCode` - Validation errors
+ * thrown either username or key are not defined.
+ *
+ */
+export function getUrl(
+	contextSpec: AmplifyServer.ContextSpec,
+	input: GetUrlInputWithPath,
+): Promise<GetUrlOutputWithPath>;
```
Comment (ashika112) [inline packages/storage/src/providers/s3/apis/server/getUrl.ts:37]: Wondering on `getUrl` if we should duplicate this for consistency or make it `GetUrlOutput` for both? Keeping together makes it just this API users in NMV.

File: packages/storage/__tests__/providers/s3/apis/copy.test.ts [hunk 2]
```diff
@@ -77,7 +82,14 @@ describe('copy API', () => {
 			afterEach(() => {
 				jest.clearAllMocks();
 			});
-			[
+			const testCases: Array<{
+				source: { accessLevel?: StorageAccessLevel; targetIdentityId?: string };
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/apis/copy.test.ts:86]: Note on this: This is how a customer would have to pass in today if it is not done directly on API call. Please let me know if that is not the case.

File: packages/storage/src/providers/s3/types/outputs.ts [hunk 3]
```diff
@@ -25,52 +26,119 @@ export interface ItemBase {
...
+ * Output type with path for S3 list API. Lists bucket objects with pagination.
+ */
+export type ListPaginateWithPathOutput =
+	StorageListOutput<ListOutputItemWithPath> & {
+		nextToken?: string;
+	};
+
+/**
+ *  Output type with path for S3 copy API.
+ * @deprecated Use {@link CopyOutputWithPath} instead.
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/types/outputs.ts:127]: Need to change this to `CopyWithPathOutput`
```suggestion
 * @deprecated Use {@link CopyWithPathOutput} instead.
```

File: packages/storage/src/providers/s3/types/outputs.ts [hunk 4]
```diff
@@ -25,52 +26,119 @@ export interface ItemBase {
...
+ */
+export type CopyOutput = Pick<ItemWithKey, 'key'>;
+/**
+ * Output type with path for S3 copy API.
+ */
+export type CopyWithPathOutput = Pick<ItemWithPath, 'path'>;
+
+/**
+ * @deprecated Use {@link RemoveOutputWithPath} instead.
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/types/outputs.ts:136]:
```suggestion
 * @deprecated Use {@link RemoveWithPathOutput} instead.
```

File: packages/storage/src/providers/s3/types/outputs.ts [hunk 5]
```diff
@@ -25,52 +26,119 @@ export interface ItemBase {
...
+ *  Output type with path for S3 uploadData API.
+ * */
+export type UploadDataWithPathOutput = UploadTask<ItemWithPath>;
 
 /**
- * Output type for S3 Copy API.
- */
-export type CopyOutput = Pick<ItemWithKeyAndPath, 'key' | 'path'>;
+ * Output type for S3 getProperties API.
+ * @deprecated Use {@link GetPropertiesOutputWithPath} instead.
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/types/outputs.ts:90]:
```suggestion
 * @deprecated Use {@link GetPropertiesWithPathOutput} instead.
```

File: packages/storage/src/providers/s3/types/outputs.ts [hunk 6]
```diff
@@ -25,52 +26,119 @@ export interface ItemBase {
...
 export type GetUrlOutput = StorageGetUrlOutput;
+/**
+ *  Output type with path for S3 getUrl API.
+ * */
+export type GetUrlWithPathOutput = StorageGetUrlOutput;
 
 /**
- * Output type for S3 getProperties API.
+ * Output type for S3 uploadData API.
+ *  @deprecated Use {@link UploadDataOutputWithPath} instead.
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/types/outputs.ts:80]:
```suggestion
 *  @deprecated Use {@link UploadDataWithPathOutput} instead.
```

File: packages/storage/src/providers/s3/types/outputs.ts [hunk 7]
```diff
@@ -25,52 +26,119 @@ export interface ItemBase {
...
 
-export type ItemWithKeyAndPath = ItemBase & StorageItem;
+/**
+ * type for S3 list item.
+ */
+export type ListOutputItem = Omit<ItemWithKey, 'metadata'>;
 
 /**
  * Output type for S3 downloadData API.
+ * @deprecated Use {@link DownloadDataOutputWithPath} instead.
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/types/outputs.ts:56]:
```suggestion
 * @deprecated Use {@link DownloadDataWithPathOutput} instead.
```

File: packages/storage/src/providers/s3/types/outputs.ts [hunk 8]
```diff
@@ -25,52 +26,119 @@ export interface ItemBase {
...
- * type for S3 list item.
+ * @deprecated Use {@link ListOutputItemWithPath} instead.
+ * type for S3 list item with key.
+ */
+export type ListOutputItemWithKey = Omit<ItemWithKey, 'metadata'>;
+
+/**
+ * type for S3 list item with path.
+ */
+export type ListOutputItemWithPath = Omit<ItemWithPath, 'metadata'>;
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/types/outputs.ts:37]: Should this be `ListItemWithPathOutput`?

File: packages/storage/src/providers/s3/types/outputs.ts [hunk 9]
```diff
@@ -25,52 +26,119 @@ export interface ItemBase {
 }
 
 /**
- * type for S3 list item.
+ * @deprecated Use {@link ListOutputItemWithPath} instead.
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/types/outputs.ts:29]: Should this be `ListItemWithPathOutput`

File: packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts [hunk 10]
```diff
@@ -218,8 +220,8 @@ export const getMultipartUploadHandlers = (
 		};
 
 		return inputType === STORAGE_INPUT_KEY
-			? { key: objectKey, path: finalKey, ...result }
-			: { path: finalKey, key: finalKey, ...result };
+			? { key: objectKey, ...result }
+			: { path: objectKey, ...result };
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts:224]: Should this be `finalKey`?

File: packages/storage/src/providers/s3/apis/server/list.ts [hunk 11]
```diff
@@ -7,86 +7,79 @@ import {
...
+export function list(
+	contextSpec: AmplifyServer.ContextSpec,
+	input: ListAllWithPathInput,
+): Promise<ListAllWithPathOutput>;
+/**
+ * @deprecated The `prefix` and `accessLevel` parameters are deprecated and may be removed in the next major version.
+ * Please use {@link https://docs.amplify.aws/react/build-a-backend/storage/list | path} instead.
+ * List files in pages with the given `prefix`.
+ * `pageSize` is defaulted to 1000. Additionally, the result will include a `nextToken` if there are more items to retrieve.
+ * @param input - The `ListPaginateInputWithPrefix` object.
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/apis/server/list.ts:50]: Should this be
```suggestion
 * @param input - The `ListPaginateWithPrefixInput` object.
```
?

File: packages/storage/src/providers/s3/apis/server/copy.ts [hunk 12]
```diff
@@ -7,57 +7,48 @@ import {
...
+): Promise<CopyWithPathOutput>;
+/**
+ * @deprecated The `key` and `accessLevel` parameters are deprecated and may be removed in the next major version.
+ * Please use {@link https://docs.amplify.aws/react/build-a-backend/storage/copy | path} instead.
+ *
+ * Copy an object from a source to a destination object within the same bucket. Can optionally copy files across
+ * different accessLevel or identityId (if source object's accessLevel is 'protected').
+ *
+ * @param contextSpec - The isolated server context.
+ * @param input - The CopyInput object.
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/apis/server/copy.ts:38]: Should this have also `CopyWithPathInput`?

File: packages/storage/src/providers/s3/apis/internal/remove.ts [hunk 13]
```diff
@@ -50,10 +55,8 @@ export const remove = async (
 	return inputType === STORAGE_INPUT_KEY
 		? {
 				key: objectKey,
-				path: finalKey,
 			}
 		: {
-				path: finalKey,
-				key: finalKey,
+				path: objectKey,
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/apis/internal/remove.ts:60]: Should this be `finalKey`?

File: packages/storage/src/providers/s3/apis/internal/getProperties.ts [hunk 14]
```diff
@@ -53,6 +58,6 @@ export const getProperties = async (
 	};
 
 	return inputType === STORAGE_INPUT_KEY
-		? { key: objectKey, path: finalKey, ...result }
-		: { path: finalKey, key: finalKey, ...result };
+		? { key: objectKey, ...result }
+		: { path: objectKey, ...result };
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/apis/internal/getProperties.ts:62]: Should this be `finalKey`?

File: packages/storage/src/providers/s3/apis/downloadData.ts [hunk 15]
```diff
@@ -153,6 +160,6 @@ const downloadDataJob =
 		};
 
 		return inputType === STORAGE_INPUT_KEY
-			? { key: objectKey, path: finalKey, ...result }
-			: { path: finalKey, key: finalKey, ...result };
+			? { key: objectKey, ...result }
+			: { path: objectKey, ...result };
```
Comment (elorzafe) [inline packages/storage/src/providers/s3/apis/downloadData.ts:164]: Should this be `finalKey`?

File: packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts (hunk 10 above)
Comment (jimblanc) [reply packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts:224]: `objectKey` should work here, functionally it's the same (controlled by the logic above on line 96)

File: packages/storage/src/providers/s3/apis/server/copy.ts (hunk 12 above)
Comment (jimblanc) [reply packages/storage/src/providers/s3/apis/server/copy.ts:38]: Not for this overload signature

**Review by jimblanc — COMMENTED** [review]
Looking good, thanks for getting this out so quickly! Variety of nits & questions below but overall looking solid.

File: packages/storage/__tests__/providers/s3/apis/copy.test.ts [hunk 16]
```diff
@@ -190,6 +196,10 @@ describe('copy API', () => {
 		});
 
 		describe('With path', () => {
+			const copyWrapper = async (
+				input: CopyWithPathInput,
+			): Promise<CopyWithPathOutput> => copy(input);
```
[the remark this answers is not in the document]
Comment (ashika112) [reply packages/storage/__tests__/providers/s3/apis/copy.test.ts:201]: So i purposefully did this so we can have both direct usage and wrapper usage. Meaning This is used in all path happy cases but error cases one will have the direct usage. I mean we want to test both ways

File: packages/storage/src/providers/s3/apis/internal/remove.ts (hunk 13 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/apis/internal/remove.ts:60]: objectKey === finalKey in term of functionality for path

File: packages/storage/src/providers/s3/apis/internal/getProperties.ts (hunk 14 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/apis/internal/getProperties.ts:62]: objectKey === finalKey in term of functionality for path

File: packages/storage/src/providers/s3/apis/downloadData.ts (hunk 15 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/apis/downloadData.ts:164]: objectKey === finalKey in term of functionality for path

**Review by erinleigh90 — CHANGES_REQUESTED** [review]
Found one issue, aside from that and the deprecation message changes pointed out by others, I'm good to approve.

File: packages/storage/src/providers/s3/types/outputs.ts (hunk 8 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/types/outputs.ts:37]: This is internal type and those still have `WithKey` `WithPath` references which we probably may clean up

File: packages/storage/src/providers/s3/types/outputs.ts (hunk 9 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/types/outputs.ts:29]: internal types still will have some refernces to `WithPath` or `WithKey`

File: packages/storage/src/providers/s3/apis/internal/list.ts [hunk 17]
```diff
@@ -7,8 +7,14 @@ import { StorageAction } from '@aws-amplify/core/internals/utils';
 import {
 	ListAllInput,
 	ListAllOutput,
+	ListAllWithPathInput,
+	ListAllWithPathOutput,
+	ListOutputItemWithKey,
```
[the remark this answers is not in the document]
Comment (ashika112) [reply packages/storage/src/providers/s3/apis/internal/list.ts:12]: I think this is miss. Thanks

File: packages/storage/src/providers/s3/apis/internal/list.ts [hunk 18]
```diff
@@ -35,8 +40,17 @@ interface ListInputArgs {
...
+	input:
+		| ListAllInput
+		| ListPaginateInput
+		| ListAllWithPathInput
+		| ListPaginateWithPathInput,
+): Promise<
+	| ListAllOutput
+	| ListPaginateOutput
+	| ListAllWithPathOutput
+	| ListAllWithPathOutput
```
[the remark this answers is not in the document]
Comment (ashika112) [reply packages/storage/src/providers/s3/apis/internal/list.ts:52]: Oh man nice catch yes 🙏

File: packages/storage/src/types/inputs.ts [hunk 19]
```diff
@@ -86,23 +86,13 @@ export interface StorageCopyInputWithKey<
...
-	source: StorageOperationInputWithPath & {
-		/** @deprecated Use path instead. */
-		key?: never;
-	};
-	destination: StorageOperationInputWithPath & {
-		/** @deprecated Use path instead. */
-		key?: never;
-	};
+	source: StorageOperationInputWithPath;
+	destination: StorageOperationInputWithPath;
```
[the remark this answers is not in the document]
Comment (ashika112) [reply packages/storage/src/types/inputs.ts:95]: there are still some internal type refernces `WithKey` and `WithPath` we do this for clear separation of concern but we should think about cleaning some of these up.

**Review by elorzafe — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Review by erinleigh90 — APPROVED** [review]

**Review by ashwinkumar6 — APPROVED** [review]

**Comment by ashwinkumar6** [comment]
lgtm, great work 🚀🚀🚀

**Pull request merged**